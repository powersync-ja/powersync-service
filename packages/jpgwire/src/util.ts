import * as tls from 'tls';
import { DEFAULT_CERTS } from './certs.js';
import * as pgwire from './pgwire.js';
import { PgType } from './pgwire_types.js';
import * as datefns from 'date-fns';

// TODO this is duplicated, but maybe that is ok
export interface NormalizedConnectionConfig {
  id: string;
  tag: string;

  hostname: string;
  port: number;
  database: string;

  username: string;
  password: string;

  sslmode: 'verify-full' | 'verify-ca' | 'disable';
  cacert: string | undefined;

  client_certificate: string | undefined;
  client_private_key: string | undefined;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export interface PgWireConnectionOptions extends NormalizedConnectionConfig {
  resolved_ip?: string;
}

export function clientTlsOptions(options: PgWireConnectionOptions): tls.ConnectionOptions {
  if (options.client_certificate && options.client_private_key) {
    return {
      cert: options.client_certificate,
      key: options.client_private_key
    };
  } else {
    return {};
  }
}
export function tlsOptions(options: PgWireConnectionOptions): false | tls.ConnectionOptions {
  if (options.sslmode == 'disable') {
    return false;
  } else if (options.sslmode == 'verify-full') {
    // Validate against default CA, or supplied CA.
    // Also validate hostname.
    return {
      ca: options.cacert ?? DEFAULT_CERTS,
      // hostname for TLS validation.
      // This may be different from the host we're connecting to if we pre-resolved
      // the IP.
      host: options.hostname,
      servername: options.hostname,
      ...clientTlsOptions(options)
    };
  } else if (options.sslmode == 'verify-ca') {
    if (options.cacert == null) {
      throw new Error(`cacert required for verify-ca`);
    }
    return {
      ca: options.cacert!,
      host: options.hostname,
      servername: options.hostname,
      // Disable hostname check
      checkServerIdentity: () => undefined,
      ...clientTlsOptions(options)
    };
  } else {
    throw new Error(`Unsupported sslmode: ${options.sslmode}`);
  }
}

export async function connectPgWire(config: PgWireConnectionOptions, options?: { type?: 'standard' | 'replication' }) {
  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password
  };

  if (options?.type != 'standard') {
    connectionOptions.replication = 'database';
  }

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }
  const connection = await pgwire.pgconnect(connectionOptions as pgwire.PgConnectOptions);
  // HACK: Replace row decoding with our own implementation
  (connection as any)._recvDataRow = _recvDataRow;
  return connection;
}

function _recvDataRow(this: any, _message: any, row: Uint8Array[], batch: any) {
  for (let i = 0; i < this._rowColumns.length; i++) {
    const valbuf = row[i];
    if (valbuf == null) {
      continue;
    }
    const { binary, typeOid } = this._rowColumns[i];
    // TODO avoid this._clientTextDecoder.decode for bytea
    row[i] = binary
      ? // do not valbuf.slice() because nodejs Buffer .slice does not copy
        // TODO but we not going to receive Buffer here ?
        Uint8Array.prototype.slice.call(valbuf)
      : PgType.decode(this._rowTextDecoder.decode(valbuf), typeOid);
  }
  batch.rows.push(row);
}

export interface PgPoolOptions {
  /**
   * Maximum number of open connections.
   *
   * Once this limit is reached, queries are queued.
   *
   * Defaults to 5.
   */
  maxSize?: number;

  /**
   * Idle timeout in ms before a connection is closed.
   */
  idleTimeout?: number | undefined;
}

/**
 * Open a postgres pool.
 *
 * The pool cannot be used for transactions, and cannot be used for logical replication.
 */
export function connectPgWirePool(config: PgWireConnectionOptions, options?: PgPoolOptions) {
  const idleTimeout = options?.idleTimeout;
  const maxSize = options?.maxSize ?? 5;

  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password,

    _poolSize: maxSize,
    _poolIdleTimeout: idleTimeout
  };

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }

  const pool = pgwire.pgpool(connectionOptions as pgwire.PgConnectOptions);
  const originalGetConnection = (pool as any)._getConnection;
  (pool as any)._getConnection = function (this: any) {
    const con = originalGetConnection.call(this);
    // HACK: Replace row decoding with our own implementation
    (con as any)._recvDataRow = _recvDataRow;
    return con;
  };
  return pool;
}

/**
 * Hack: sslrootcert is passed through as-is to pgwire_node.
 *
 * We use that to pass in custom TLS options, without having to modify pgwire itself.
 */
export interface ExtendedPgwireOptions extends pgwire.PgConnectKnownOptions {
  sslrootcert?: false | tls.ConnectionOptions;
}

export function lsnMakeComparable(text: string) {
  const [h, l] = text.split('/');
  return h.padStart(8, '0') + '/' + l.padStart(8, '0');
}

/**
 * Convert a postgres timestamptz to a SQLite-compatible/normalized timestamp.
 *
 * We output in the format:
 * YYYY-MM-DD HH:MM:SSZ
 * YYYY-MM-DD HH:MM:SS.SSSZ
 *
 * Precision is preserved from the source string, with the exception of ".000".
 *
 * We have specific exceptions for -infinity and infinity.
 */
export function timestamptzToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  // Make compatible with SQLite
  const match = /^([\d\-]+) ([\d:]+)(\.\d+)?([+-][\d:]+)$/.exec(source);
  if (match == null) {
    if (source == 'infinity') {
      return '9999-12-31 23:59:59Z';
    } else if (source == '-infinity') {
      return '0000-01-01 00:00:00Z';
    } else {
      return null;
    }
  }
  const [_, date, time, precision, timezone] = match as any;
  // datefns.parseISO() supports some dates that new Date() doesn't, e.g. '0022-09-08 21:19:28+00'
  const parsed = datefns.parseISO(`${date}T${time}${timezone}`);
  // Some dates are outside the range of JS and/or SQLite, e.g. 10022-09-08 21:19:28+00.
  // Just fallback to null for now.
  // We may consider preserving the source value in these cases.
  if (isNaN(parsed.getTime())) {
    return null;
  }
  const text = parsed.toISOString().replace('T', ' ').replace('.000', '').replace('Z', '');

  return `${text}${precision ?? ''}Z`;
}

/**
 * For timestamp without timezone, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function timestampToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31 23:59:59';
  } else if (source == '-infinity') {
    return '0000-01-01 00:00:00';
  } else {
    return source;
  }
}
/**
 * For date, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function dateToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31';
  } else if (source == '-infinity') {
    return '0000-01-01';
  } else {
    return source;
  }
}

/**
 * PgResult rows are arrays of values.
 *
 * This converts it to objects.
 */
export function pgwireRows(rs: pgwire.PgResult): Record<string, any>[] {
  const columns = rs.columns;
  return rs.rows.map((row) => {
    let r: Record<string, any> = {};
    for (let i = 0; i < columns.length; i++) {
      const c = columns[i];
      r[c.name] = row[i];
    }
    return r;
  });
}
