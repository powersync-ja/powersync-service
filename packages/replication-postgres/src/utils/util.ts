import * as pgwire from '@powersync/service-jpgwire';
import { replication } from '@powersync/service-core';

import { ReplicaIdentityResult } from '../types.js';
import { retriedQuery } from './pgwire_utils.js';

export function getReplicaIdColumns(relation: pgwire.PgoutputRelation): replication.ReplicationColumn[] {
  if (relation.replicaIdentity == 'nothing') {
    return [];
  } else {
    return relation.columns.filter((c) => (c.flags & 0b1) != 0).map((c) => ({ name: c.name, typeOid: c.typeOid }));
  }
}

export function getRelId(source: pgwire.PgoutputRelation): number {
  // Source types are wrong here
  const relId = (source as any).relationOid as number;
  if (relId == null || typeof relId != 'number') {
    throw new Error(`No relation id!`);
  }
  return relId;
}

export function getPgOutputRelation(source: pgwire.PgoutputRelation): replication.Relation {
  return {
    name: source.name,
    schema: source.schema,
    relationId: getRelId(source),
    replicaIdentity: source.replicaIdentity,
    replicationColumns: getReplicaIdColumns(source)
  };
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<replication.ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<replication.ReplicationColumn[]> {
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
