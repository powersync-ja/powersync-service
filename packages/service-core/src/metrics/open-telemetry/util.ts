import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { MeterProvider, MetricReader, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { logger } from '@powersync/lib-services-framework';
import { ServiceContext } from '../../system/ServiceContext.js';
import { MetricsFactory } from '../metrics-interfaces.js';
import { OpenTelemetryMetricsFactory } from './OpenTelemetryMetricsFactory.js';

import pkg from '../../../package.json' with { type: 'json' };
import { resourceFromAttributes } from '@opentelemetry/resources';

export function createOpenTelemetryMetricsFactory(context: ServiceContext): MetricsFactory {
  const { configuration, lifeCycleEngine, storageEngine } = context;
  const configuredExporters: MetricReader[] = [];

  if (configuration.telemetry.prometheus_port) {
    const prometheusExporter = new PrometheusExporter({
      port: configuration.telemetry.prometheus_port,
      preventServerStart: true
    });
    configuredExporters.push(prometheusExporter);

    lifeCycleEngine.withLifecycle(prometheusExporter, {
      start: async () => {
        await prometheusExporter.startServer();
        logger.info(`Prometheus metric export enabled on port:${configuration.telemetry.prometheus_port}`);
      }
    });
  }

  if (!configuration.telemetry.disable_telemetry_sharing) {
    const periodicExporter = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: configuration.telemetry.internal_service_endpoint
      }),
      exportIntervalMillis: 1000 * 60 * 5 // 5 minutes
    });

    configuredExporters.push(periodicExporter);
  }

  let resolvedInstanceId: (id: string) => void;
  const instanceIdPromise = new Promise<string>((resolve) => {
    resolvedInstanceId = resolve;
  });

  lifeCycleEngine.withLifecycle(null, {
    start: async () => {
      const bucketStorage = storageEngine.activeBucketStorage;
      try {
        const instanceId = await bucketStorage.getPowerSyncInstanceId();
        resolvedInstanceId(instanceId);
      } catch (err) {
        resolvedInstanceId('Unknown');
      }
    }
  });

  const resource = resourceFromAttributes({
    ['service']: 'PowerSync',
    ['service.version']: pkg.version,
    ['instance_id']: instanceIdPromise
  });

  // This triggers OpenTelemetry to resolve the async attributes (instanceIdPromise).
  // This will never reject, and we don't specifically need to wait for it.
  resource.waitForAsyncAttributes?.();

  const meterProvider = new MeterProvider({
    resource,

    readers: configuredExporters
  });

  lifeCycleEngine.withLifecycle(meterProvider, {
    stop: async () => {
      await meterProvider.shutdown();
    }
  });

  const meter = meterProvider.getMeter('powersync');

  return new OpenTelemetryMetricsFactory(meter);
}
