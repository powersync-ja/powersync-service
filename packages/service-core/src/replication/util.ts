import * as pgwire from '@powersync/service-jpgwire';

import * as util from '../util/util-index.js';
import { ReplicationColumn, ReplicationIdentity } from './PgRelation.js';

export interface ReplicaIdentityResult {
  columns: ReplicationColumn[];
  replicationIdentity: ReplicationIdentity;
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<ReplicationColumn[]> {
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await util.retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
