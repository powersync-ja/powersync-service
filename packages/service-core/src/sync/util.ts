import * as timers from 'timers/promises';

import * as util from '../util/util-index.js';
import { Metrics } from '../metrics/Metrics.js';

export type TokenStreamOptions = {
  /**
   * Adds periodic keepalive events
   */
  keep_alive: boolean;
  /**
   * Warn before the token is going to expire
   */
  expire_warning_period: number;
};

const KEEPALIVE_INTERVAL = 20_000;

const DEFAULT_TOKEN_STREAM_OPTIONS: TokenStreamOptions = {
  keep_alive: true,
  expire_warning_period: 20_000
};

/**
 * An iterator that periodically yields token and optionally keepalive events, and returns once the
 * provided token expiry is reached.
 *
 * @param token exp is expiry as a unix timestamp (seconds)
 * @param signal abort the iterator with this
 * @param options configure keepalive and expire warnings
 */
export async function* tokenStream(
  token: { exp: number },
  signal: AbortSignal,
  options?: Partial<TokenStreamOptions>
): AsyncGenerator<util.StreamingSyncKeepalive> {
  const resolved_options: TokenStreamOptions = {
    ...DEFAULT_TOKEN_STREAM_OPTIONS,
    ...(options ?? {})
  };

  const { keep_alive, expire_warning_period } = resolved_options;

  // Real tokens always have an integer for exp.
  // In tests, we may use fractional seconds to reduce the delay.
  // Both cases are handled here.
  const expires_at = token.exp * 1000;
  const expire_warning_at = expires_at - expire_warning_period;

  let first_expire_test = true;

  while (!signal.aborted) {
    const token_expires_in = Math.max(0, Math.ceil(token.exp - Date.now() / 1000));
    if (first_expire_test && token_expires_in > 0) {
      first_expire_test = false;
    } else {
      yield { token_expires_in: token_expires_in };
      if (token_expires_in == 0) {
        return;
      }
    }

    const keep_alive_delay = KEEPALIVE_INTERVAL * (0.9 + Math.random() * 0.2);

    // Add margin due to setTimeout inaccuracies
    const expiry_delay = Math.max(0, expires_at - Date.now() + 3);
    const expiry_warning_delay = Math.max(0, expire_warning_at - Date.now() + 3);
    // Either the warning has past or it's before
    const relevant_expiry_delay = expiry_warning_delay != 0 ? expiry_warning_delay : expiry_delay;

    const delay = keep_alive ? Math.min(relevant_expiry_delay, keep_alive_delay) : relevant_expiry_delay;
    await timers.setTimeout(delay, null, { signal }).catch(() => {
      // Ignore AbortError
    });
  }
}

export async function* ndjson(iterator: AsyncIterable<string | null | Record<string, any>>): AsyncGenerator<string> {
  for await (let data of iterator) {
    if (data == null) {
      // Empty value to flush iterator memory
      continue;
    } else if (typeof data == 'string') {
      // Pre-serialized value
      yield data + '\n';
    } else {
      yield JSON.stringify(data) + '\n';
    }
  }
}

export async function* transformToBytesTracked(iterator: AsyncIterable<string>): AsyncGenerator<Buffer> {
  for await (let data of iterator) {
    const encoded = Buffer.from(data, 'utf8');
    Metrics.getInstance().data_synced_bytes.add(encoded.length);
    yield encoded;
  }
}
