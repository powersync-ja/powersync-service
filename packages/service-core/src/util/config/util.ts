import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
