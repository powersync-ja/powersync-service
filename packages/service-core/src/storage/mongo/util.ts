import { SqliteJsonValue } from '@powersync/service-sync-rules';
import * as bson from 'bson';
import * as crypto from 'crypto';
import * as mongo from 'mongodb';
import * as uuid from 'uuid';
import { OplogEntry } from '../../util/protocol-types.js';
import { ID_NAMESPACE, timestampToOpId } from '../../util/utils.js';
import { BucketDataDocument, ReplicaId } from './models.js';

/**
 * Lookup serialization must be number-agnostic. I.e. normalize numbers, instead of preserving numbers.
 * @param lookup
 */

export function serializeLookup(lookup: SqliteJsonValue[]) {
  const normalized = lookup.map((value) => {
    if (typeof value == 'number' && Number.isInteger(value)) {
      return BigInt(value);
    } else {
      return value;
    }
  });
  return new bson.Binary(bson.serialize({ l: normalized }));
}

export function idPrefixFilter<T>(prefix: Partial<T>, rest: (keyof T)[]): mongo.Condition<T> {
  let filter = {
    $gte: {
      ...prefix
    } as any,
    $lt: {
      ...prefix
    } as any
  };

  for (let key of rest) {
    filter.$gte[key] = new bson.MinKey();
    filter.$lt[key] = new bson.MaxKey();
  }

  return filter;
}

export function generateSlotName(prefix: string, sync_rules_id: number) {
  const slot_suffix = crypto.randomBytes(2).toString('hex');
  return `${prefix}${sync_rules_id}_${slot_suffix}`;
}

/**
 * Read a single batch of data from a cursor, then close it.
 *
 * We do our best to avoid MongoDB fetching any more data than this single batch.
 *
 * This is similar to using `singleBatch: true` in find options.
 * However, that makes `has_more` detection very difficult, since the cursor is always closed
 * after the first batch. Instead, we do a workaround to only fetch a single batch below.
 *
 * For this to be effective, set batchSize = limit in the find command.
 */
export async function readSingleBatch<T>(cursor: mongo.FindCursor<T>): Promise<{ data: T[]; hasMore: boolean }> {
  try {
    let data: T[];
    let hasMore = true;
    // Let MongoDB load the first batch of data
    const hasAny = await cursor.hasNext();
    // Now it's in memory, and we can read it
    data = cursor.readBufferedDocuments();
    if (!hasAny || cursor.id?.isZero()) {
      // A zero id means the cursor is exhaused.
      // No results (hasAny == false) means even this batch doesn't have data.
      // This should similar results as `await cursor.hasNext()`, but without
      // actually fetching the next batch.
      //
      // Note that it is safe (but slightly inefficient) to return `hasMore: true`
      // without there being more data, as long as the next batch
      // will return `hasMore: false`.
      hasMore = false;
    }
    return { data, hasMore };
  } finally {
    // Match the from the cursor iterator logic here:
    // https://github.com/mongodb/node-mongodb-native/blob/e02534e7d1c627bf50b85ca39f5995dbf165ad44/src/cursor/abstract_cursor.ts#L327-L331
    if (!cursor.closed) {
      await cursor.close();
    }
  }
}

export const BSON_DESERIALIZE_OPTIONS: bson.DeserializeOptions = {
  // use bigint instead of Long
  useBigInt64: true
};

export function mapOpEntry(row: BucketDataDocument): OplogEntry {
  if (row.op == 'PUT' || row.op == 'REMOVE') {
    return {
      op_id: timestampToOpId(row._id.o),
      op: row.op,
      object_type: row.table,
      object_id: row.row_id,
      checksum: Number(row.checksum),
      subkey: replicaIdToSubkey(row.source_table!, row.source_key!),
      data: row.data
    };
  } else {
    // MOVE, CLEAR

    return {
      op_id: timestampToOpId(row._id.o),
      op: row.op,
      checksum: Number(row.checksum)
    };
  }
}

/**
 * Returns true if two ReplicaId values are the same (serializes to the same BSON value).
 */
export function replicaIdEquals(a: ReplicaId, b: ReplicaId) {
  if (a === b) {
    return true;
  } else if (typeof a == 'string' && typeof b == 'string') {
    return a == b;
  } else if (isUUID(a) && isUUID(b)) {
    return a.equals(b);
  } else if (a == null && b == null) {
    return true;
  } else if (a != null || b != null) {
    return false;
  } else {
    // There are many possible primitive values, this covers them all
    return (bson.serialize({ id: a }) as Buffer).equals(bson.serialize({ id: b }));
  }
}

export function replicaIdToSubkey(table: bson.ObjectId, id: ReplicaId): string {
  if (isUUID(id)) {
    // Special case for UUID for backwards-compatiblity
    return `${table.toHexString()}/${id.toHexString()}`;
  } else {
    // Hashed UUID from the table and id
    const repr = bson.serialize({ table, id });
    return uuid.v5(repr, ID_NAMESPACE);
  }
}

/**
 * True if this is a bson.UUID.
 *
 * Works even with multiple copies of the bson package.
 */
export function isUUID(value: any): value is bson.UUID {
  if (value == null || typeof value != 'object') {
    return false;
  }
  const uuid = value as bson.UUID;
  return uuid._bsontype == 'Binary' && uuid.sub_type == bson.Binary.SUBTYPE_UUID;
}
