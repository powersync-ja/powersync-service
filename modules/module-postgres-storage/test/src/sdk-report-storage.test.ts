import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { POSTGRES_REPORT_STORAGE_FACTORY } from './util.js';

describe('SDK reporting storage', () => {
  const now = new Date();
  const nowAdd5minutes = new Date(
    now.getFullYear(),
    now.getMonth(),
    now.getDate(),
    now.getHours(),
    now.getMinutes() + 5
  );
  const nowLess5minutes = new Date(
    now.getFullYear(),
    now.getMonth(),
    now.getDate(),
    now.getHours(),
    now.getMinutes() - 5
  );
  const yesterday = new Date(now.getFullYear(), now.getMonth(), now.getDate() - 1);
  const user_one = {
    user_id: 'user_one',
    client_id: 'client_one',
    connect_at: now.toISOString(),
    sdk: 'powersync-dart/1.6.4',
    user_agent: 'powersync-dart/1.6.4 Dart (flutter-web) Chrome/128 android',
    jwt_exp: nowAdd5minutes.toISOString(),
    id: '1'
  };
  const user_two = {
    user_id: 'user_two',
    client_id: 'client_two',
    connect_at: nowLess5minutes.toISOString(),
    sdk: 'powersync-js/1.21.0',
    user_agent: 'powersync-js/1.21.0 powersync-web Chromium/138 linux',
    jwt_exp: nowAdd5minutes.toISOString(),
    id: '2'
  };
  const user_three = {
    user_id: 'user_three',
    client_id: 'client_three',
    connect_at: yesterday.toISOString(),
    sdk: 'powersync-js/1.21.0',
    user_agent: 'powersync-js/1.21.0 powersync-web Firefox/141 linux',
    disconnect_at: yesterday.toISOString(),
    id: '3'
  };

  const user_four = {
    user_id: 'user_four',
    client_id: 'user_five',
    connect_at: now.toISOString(),
    sdk: 'powersync-js/1.21.0',
    user_agent: 'powersync-js/1.21.0 powersync-web Firefox/141 linux',
    jwt_exp: nowLess5minutes.toISOString(),
    id: '4'
  };

  beforeAll(async () => {
    const { db } = await POSTGRES_REPORT_STORAGE_FACTORY();
    const result = await db.sql`
      INSERT INTO
        sdk_report_events (
          user_id,
          client_id,
          connect_at,
          sdk,
          user_agent,
          jwt_exp,
          id,
          disconnect_at
        )
      VALUES
        (
          ${{ type: 'varchar', value: user_one.user_id }},
          ${{ type: 'varchar', value: user_one.client_id }},
          ${{ type: 1184, value: user_one.connect_at }},
          ${{ type: 'varchar', value: user_one.sdk }},
          ${{ type: 'varchar', value: user_one.user_agent }},
          ${{ type: 1184, value: user_one.jwt_exp }},
          ${{ type: 'varchar', value: user_one.id }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: user_two.user_id }},
          ${{ type: 'varchar', value: user_two.client_id }},
          ${{ type: 1184, value: user_two.connect_at }},
          ${{ type: 'varchar', value: user_two.sdk }},
          ${{ type: 'varchar', value: user_two.user_agent }},
          ${{ type: 1184, value: user_two.jwt_exp }},
          ${{ type: 'varchar', value: user_two.id }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: user_four.user_id }},
          ${{ type: 'varchar', value: user_four.client_id }},
          ${{ type: 1184, value: user_four.connect_at }},
          ${{ type: 'varchar', value: user_four.sdk }},
          ${{ type: 'varchar', value: user_four.user_agent }},
          ${{ type: 1184, value: user_four.jwt_exp }},
          ${{ type: 'varchar', value: user_four.id }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: user_three.user_id }},
          ${{ type: 'varchar', value: user_three.client_id }},
          ${{ type: 1184, value: user_three.connect_at }},
          ${{ type: 'varchar', value: user_three.sdk }},
          ${{ type: 'varchar', value: user_three.user_agent }},
          NULL,
          ${{ type: 'varchar', value: user_three.id }},
          ${{ type: 1184, value: user_three.disconnect_at }}
        )
    `.execute();
    console.log(result);
  });
  afterAll(async () => {
    const { db } = await POSTGRES_REPORT_STORAGE_FACTORY();
    await db.sql`DELETE FROM sdk_report_events`.execute();
  });
  it('Should show currently connected users with start range', async () => {
    const factory = await POSTGRES_REPORT_STORAGE_FACTORY();
    const current = await factory.listCurrentConnections({
      range: {
        start_date: new Date(
          now.getFullYear(),
          now.getMonth(),
          now.getDate(),
          now.getHours(),
          now.getMinutes() - 1
        ).toISOString()
      }
    });
    expect(current).toMatchSnapshot();
  });
  it('Should show currently connected users with start range and end range', async () => {
    const factory = await POSTGRES_REPORT_STORAGE_FACTORY();
    const current = await factory.listCurrentConnections({
      range: {
        end_date: nowLess5minutes.toISOString(),
        start_date: new Date(
          now.getFullYear(),
          now.getMonth(),
          now.getDate(),
          now.getHours(),
          now.getMinutes() - 6
        ).toISOString()
      }
    });
    expect(current).toMatchSnapshot();
  });
});
