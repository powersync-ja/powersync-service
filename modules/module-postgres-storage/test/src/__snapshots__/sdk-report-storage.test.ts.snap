// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SDK reporting storage > Should show currently connected users with start range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
  ],
  "users": 1,
}
`;

exports[`SDK reporting storage > Should show currently connected users with start range and end range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.0",
      "users": 1,
    },
  ],
  "users": 1,
}
`;
